### PR TITLE
feat(gatsby): improve refresh endpoint output

### PR DESCRIPTION
**Pull request by wardpeet**
## Description

Add status-code and extra information when it's not available. This will help us adopt the /__refresh/<pluginname> endpoint. This makes it easier to feature detect the specific endpoint.

### CODE REVIEW

File: packages/gatsby/src/utils/start-server.ts [hunk 1]
```diff
@@ -259,6 +259,14 @@ module.exports = {
 
     if (enableRefresh && authorizedRefresh) {
       refresh(req, pluginName)
+      res.status(200)
+      res.setHeader(`content-type`, `application/json`)
+    } else {
+      res.status(404)
```
Comment (ascorbic) [inline packages/gatsby/src/utils/start-server.ts:265]: Is it worth distinguishing between disabled and unauthorised with a 403?

**Review by ascorbic — APPROVED** [review]

File: packages/gatsby/src/utils/start-server.ts (hunk 1 above)
Comment (wardpeet) [reply packages/gatsby/src/utils/start-server.ts:265]: For our use case not but doesn't harm us

File: packages/gatsby/src/utils/start-server.ts [hunk 2]
```diff
@@ -469,6 +469,12 @@ module.exports = {
   }
 
   app.use(async (req, res) => {
+    // in this catch-all block we don't support POST so we should 404
+    if (req.method === `POST`) {
+      res.status(404).end()
```
Comment (ascorbic) [inline packages/gatsby/src/utils/start-server.ts:474]: To be an HTTP pedant again (and I'm sure it makes no difference) but this should be `405`! 😄

**Review by ascorbic — APPROVED** [review]

File: packages/gatsby/src/utils/start-server.ts (hunk 2 above)
Comment (wardpeet) [reply packages/gatsby/src/utils/start-server.ts:474]: I don't know, this sounds like a 404. For me 405 (and I could be wrong) is when we don't support POST at all?

File: packages/gatsby/src/utils/start-server.ts [hunk 3]
```diff
@@ -259,6 +259,14 @@ module.exports = {
 
     if (enableRefresh && authorizedRefresh) {
       refresh(req, pluginName)
+      res.status(200)
+      res.setHeader(`content-type`, `application/json`)
+    } else {
+      res.status(authorizedRefresh ? 404 : 403)
+      res.json({
+        error: `Refresh failed`,
```
Comment (pieh) [inline packages/gatsby/src/utils/start-server.ts:267]: If you want to make it easier to adopt - I'd suggest expanding on error message (stuff below will likely need to use prettier if applied):
```suggestion
        error: enableRefresh ? `Authorization failed. Make sure you add authorization header to your refresh requests` : `Refresh endpoint is not enabled. Run gatsby with ENABLE_GATSBY_REFRESH_ENDPOINT=true environtment variable set.`,
```

File: packages/gatsby/src/utils/start-server.ts (hunk 2 above)
Comment (ascorbic) [reply packages/gatsby/src/utils/start-server.ts:474]: Maybe. It's "Method Not Allowed", but I can see that the description could be read either way. In any case, not a big deal

**Review by pieh — APPROVED** [review]
Thank you ❤️

File: packages/gatsby/src/utils/start-server.ts (hunk 2 above)
Comment (pieh) [reply packages/gatsby/src/utils/start-server.ts:476]: I checked what GitHub does for unspported methods and they also return 404 status (i.e. https://api.github.com/repos/gatsbyjs/gatsby/commits/1e702ae64e08e8b5d22ca448d870c9bb24954bf8 with GET is response, but with PATCH is 404)

Comment (pieh) [reply packages/gatsby/src/utils/start-server.ts:476]: so 404 is fine IMO

**Pull request merged**